File: Se7enPro/Services/CoreUpdateService.cs
using System;
using System.Diagnostics;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;

namespace Se7enPro.Services;

public sealed class CoreUpdateService : ICoreUpdateService, IDisposable
{
    private const string AetherAssetName = "aether-windows-x86_64.zip";
    private const string ReleaseApiUrl =
        "https://api.github.com/repos/CluvexStudio/Aether/releases/latest";

    private const string ReleaseDownloadPrefix =
        "https://github.com/CluvexStudio/Aether/releases/download/";

    private const string RequiredSignerSubject = "Cluvex";

    private static readonly bool RequireSignedCoreBinaries = true;

    private const long MaxArchiveBytes = 128L * 1024 * 1024;

    private const long MaxExtractedBytes = 256L * 1024 * 1024;

    private readonly ILogger<CoreUpdateService> _logger;
    private readonly HttpClient _httpClient;

    public CoreUpdateService(ILogger<CoreUpdateService> logger)
    {
        _logger = logger;
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };
        _httpClient.DefaultRequestHeaders.UserAgent.ParseAdd("Se7enPro-CoreUpdater/1.0");
    }

    public void Dispose() => _httpClient.Dispose();

    private static bool IsTrustedDownloadUrl(string url)
    {
        if (!Uri.TryCreate(url, UriKind.Absolute, out var uri)) return false;
        if (!uri.Scheme.Equals(Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase)) return false;
        if (!uri.Host.Equals("github.com", StringComparison.OrdinalIgnoreCase)) return false;
        if (!string.IsNullOrEmpty(uri.UserInfo)) return false;
        return uri.AbsoluteUri.StartsWith(ReleaseDownloadPrefix, StringComparison.Ordinal);
    }

    public const string UnknownVersion = "unknown";

    public string GetInstalledVersion(string coreId)
    {
        try
        {
            switch (coreId.ToLowerInvariant())
            {
                case "aether":
                {
                    var exePath = FindAetherExecutable();
                    if (!string.IsNullOrEmpty(exePath) && File.Exists(exePath))
                    {
                        var ver = QueryExeVersion(exePath, "--version");
                        if (!string.IsNullOrEmpty(ver))
                        {

                            var parts = ver.Split(' ', StringSplitOptions.RemoveEmptyEntries);
                            if (parts.Length >= 2) return parts[1].Trim();
                            return ver.Trim();
                        }

                        var fvi = FileVersionInfo.GetVersionInfo(exePath);
                        if (!string.IsNullOrEmpty(fvi.FileVersion))
                            return fvi.FileVersion;
                    }
                    return UnknownVersion;
                }

                case "tor":
                {
                    var exePath = FindTorExecutable();
                    if (!string.IsNullOrEmpty(exePath) && File.Exists(exePath))
                    {
                        var ver = QueryExeVersion(exePath, "--version");
                        if (!string.IsNullOrEmpty(ver))
                        {

                            var lines = ver.Split('\n');
                            if (lines.Length > 0)
                            {
                                var first = lines[0].Trim();
                                var idx = first.IndexOf("version ", StringComparison.OrdinalIgnoreCase);
                                if (idx >= 0)
                                {
                                    var rest = first.Substring(idx + 8).Trim();
                                    var space = rest.IndexOf(' ');
                                    return space > 0 ? rest.Substring(0, space) : rest;
                                }
                            }
                        }

                        try
                        {
                            var fvi = FileVersionInfo.GetVersionInfo(exePath);
                            if (!string.IsNullOrEmpty(fvi.FileVersion)) return fvi.FileVersion;
                        }
                        catch { }
                    }
                    return UnknownVersion;
                }

                default:
                    return UnknownVersion;
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to determine installed version for core: {CoreId}", coreId);
            return UnknownVersion;
        }
    }

    public async Task<CoreUpdateInfo> CheckForUpdateAsync(string coreId, CancellationToken ct = default)
    {
        var installed = GetInstalledVersion(coreId);

        if (!coreId.Equals("aether", StringComparison.OrdinalIgnoreCase))
        {

            return new CoreUpdateInfo(
                CoreId: coreId,
                DisplayName: GetCoreDisplayName(coreId),
                InstalledVersion: installed,
                LatestVersion: UnknownVersion,
                HasUpdate: false,
                DownloadUrl: "",
                ReleaseNotes: "This engine ships with the app; Se7en does not check a remote "
                              + "release feed for it, so no update status is available.",
                DownloadSizeBytes: 0
            );
        }

        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, ReleaseApiUrl);
            using var res = await _httpClient.SendAsync(req, ct);
            res.EnsureSuccessStatusCode();

            var json = await res.Content.ReadAsStringAsync(ct);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            var tagName = root.TryGetProperty("tag_name", out var tagElem) ? tagElem.GetString() ?? "" : "";
            var releaseNotes = root.TryGetProperty("body", out var bodyElem) ? bodyElem.GetString() ?? "" : "";

            var latestVersion = tagName.TrimStart('v', 'V').Trim();
            var downloadUrl = "";
            long sizeBytes = 0;

            if (root.TryGetProperty("assets", out var assetsElem) && assetsElem.ValueKind == JsonValueKind.Array)
            {
                foreach (var asset in assetsElem.EnumerateArray())
                {
                    var name = asset.TryGetProperty("name", out var nameElem) ? nameElem.GetString() ?? "" : "";
                    if (name.Equals(AetherAssetName, StringComparison.OrdinalIgnoreCase))
                    {
                        downloadUrl = asset.TryGetProperty("browser_download_url", out var urlElem) ? urlElem.GetString() ?? "" : "";
                        sizeBytes = asset.TryGetProperty("size", out var sizeElem) ? sizeElem.GetInt64() : 0;
                        break;
                    }
                }
            }

            if (string.IsNullOrEmpty(downloadUrl) && !string.IsNullOrEmpty(tagName))
            {
                downloadUrl = $"{ReleaseDownloadPrefix}{Uri.EscapeDataString(tagName)}/{AetherAssetName}";
            }

            if (!string.IsNullOrEmpty(downloadUrl) && !IsTrustedDownloadUrl(downloadUrl))
            {
                _logger.LogWarning(
                    "Rejecting Aether asset URL outside the expected release path: {Url}", downloadUrl);
                downloadUrl = "";
            }

            var hasUpdate = IsNewerVersion(installed, latestVersion);

            return new CoreUpdateInfo(
                CoreId: coreId,
                DisplayName: "Aether (WARP / MASQUE)",
                InstalledVersion: installed,
                LatestVersion: latestVersion,
                HasUpdate: hasUpdate,
                DownloadUrl: downloadUrl,
                ReleaseNotes: releaseNotes,
                DownloadSizeBytes: sizeBytes
            );
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to check for updates for core {CoreId}", coreId);
            throw;
        }
    }

    public async Task<bool> UpdateCoreAsync(string coreId, IProgress<int>? progress = null, CancellationToken ct = default)
    {
        if (!coreId.Equals("aether", StringComparison.OrdinalIgnoreCase))
        {
            throw new NotSupportedException($"Updating core '{coreId}' is not supported yet.");
        }

        var updateInfo = await CheckForUpdateAsync(coreId, ct);
        if (string.IsNullOrEmpty(updateInfo.DownloadUrl))
        {
            throw new InvalidOperationException("Could not find download URL for Aether update.");
        }

        if (!IsTrustedDownloadUrl(updateInfo.DownloadUrl))
        {
            throw new InvalidOperationException(
                $"Refusing to download the core from an untrusted URL: {updateInfo.DownloadUrl}");
        }

        var tempDir = Path.Combine(Path.GetTempPath(), "Se7enPro_Update_" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(tempDir);
        var zipPath = Path.Combine(tempDir, AetherAssetName);
        var extractedExe = Path.Combine(tempDir, "aether.exe");

        try
        {

            progress?.Report(5);
            await DownloadArchiveAsync(updateInfo, zipPath, progress, ct);

            progress?.Report(72);

            ExtractAetherExe(zipPath, extractedExe);

            progress?.Report(78);

            var trust = BinaryTrust.VerifyAuthenticode(extractedExe, RequiredSignerSubject);
            if (!trust.Trusted)
            {
                if (RequireSignedCoreBinaries)
                {
                    throw new InvalidOperationException(
                        $"The downloaded Aether binary failed signature verification and was NOT installed: "
                        + $"{trust.Detail}");
                }
                _logger.LogWarning(
                    "SECURITY: installing an unverified Aether binary because signature enforcement is "
                    + "disabled in this build: {Detail}", trust.Detail);
            }
            else
            {
                _logger.LogInformation("Downloaded Aether binary verified: {Detail}", trust.Detail);
            }

            progress?.Report(84);

            KillOwnAetherProcesses();

            progress?.Report(90);

            var localAppCachedPath = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "Se7en", "aether", EngineProcessNames.Aether);
            AtomicFile.Install(extractedExe, localAppCachedPath);

            var appResourcePath = Path.Combine(AppContext.BaseDirectory, "Resources", "aether", "aether.exe");
            try
            {
                AtomicFile.Install(extractedExe, appResourcePath);
            }
            catch (Exception ex)
            {

                _logger.LogInformation(ex,
                    "Could not refresh the bundled copy at {Path}; the cached core was updated", appResourcePath);
            }

            progress?.Report(96);

            var verifiedVer = QueryExeVersion(localAppCachedPath, "--version");
            if (string.IsNullOrWhiteSpace(verifiedVer))
            {
                throw new InvalidOperationException(
                    "The core was installed but did not respond to --version; the previous binary has been "
                    + "left in place as .bak next to it.");
            }

            progress?.Report(100);
            _logger.LogInformation("Aether core successfully installed/verified: {Version}", verifiedVer);
            return true;
        }
        finally
        {
            try
            {
                if (Directory.Exists(tempDir))
                    Directory.Delete(tempDir, recursive: true);
            }
            catch { }
        }
    }

    private async Task DownloadArchiveAsync(
        CoreUpdateInfo updateInfo, string zipPath, IProgress<int>? progress, CancellationToken ct)
    {
        using var response = await _httpClient.GetAsync(
            updateInfo.DownloadUrl, HttpCompletionOption.ResponseHeadersRead, ct);
        response.EnsureSuccessStatusCode();

        var finalUrl = response.RequestMessage?.RequestUri?.AbsoluteUri;
        if (!string.IsNullOrEmpty(finalUrl) && !IsTrustedDownloadUrl(finalUrl))
        {

            if (!finalUrl.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
            {
                throw new InvalidOperationException(
                    $"The core download was redirected to a non-HTTPS location: {finalUrl}");
            }
        }

        var declared = response.Content.Headers.ContentLength
                       ?? (updateInfo.DownloadSizeBytes > 0 ? updateInfo.DownloadSizeBytes : 0);
        if (declared > MaxArchiveBytes)
        {
            throw new InvalidOperationException(
                $"The core archive declares {declared:N0} bytes, above the {MaxArchiveBytes:N0} byte limit.");
        }

        await using var stream = await response.Content.ReadAsStreamAsync(ct);
        await using var fileStream = new FileStream(
            zipPath, FileMode.Create, FileAccess.Write, FileShare.None, 81920, useAsync: true);

        var buffer = new byte[81920];
        long downloadedBytes = 0;
        int bytesRead;

        while ((bytesRead = await stream.ReadAsync(buffer, ct)) > 0)
        {
            downloadedBytes += bytesRead;
            if (downloadedBytes > MaxArchiveBytes)
            {
                throw new InvalidOperationException(
                    $"The core archive exceeded the {MaxArchiveBytes:N0} byte limit mid-download.");
            }
            await fileStream.WriteAsync(buffer.AsMemory(0, bytesRead), ct);

            if (declared > 0)
            {
                var pct = (int)(5 + (downloadedBytes * 65 / declared));
                progress?.Report(Math.Min(70, pct));
            }
        }
    }

    private static void ExtractAetherExe(string zipPath, string destination)
    {
        using var archive = ZipFile.OpenRead(zipPath);
        var entry = archive.GetEntry("aether.exe") ??
                    archive.Entries.FirstOrDefault(
                        e => e.Name.Equals("aether.exe", StringComparison.OrdinalIgnoreCase));

        if (entry is null)
        {
            throw new FileNotFoundException("aether.exe was not found inside the downloaded archive.");
        }
        if (entry.Length > MaxExtractedBytes)
        {
            throw new InvalidOperationException(
                $"aether.exe inside the archive declares {entry.Length:N0} bytes, above the "
                + $"{MaxExtractedBytes:N0} byte limit.");
        }

        using var source = entry.Open();
        using var target = new FileStream(destination, FileMode.Create, FileAccess.Write, FileShare.None);
        var buffer = new byte[81920];
        long written = 0;
        int read;
        while ((read = source.Read(buffer, 0, buffer.Length)) > 0)
        {
            written += read;
            if (written > MaxExtractedBytes)
            {
                throw new InvalidOperationException(
                    $"aether.exe expanded past the {MaxExtractedBytes:N0} byte limit.");
            }
            target.Write(buffer, 0, read);
        }
    }

    private void KillOwnAetherProcesses()
    {
        var ourImages = new[]
        {
            Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "Se7en", "aether", EngineProcessNames.Aether),
            Path.Combine(AppContext.BaseDirectory, "Resources", "aether", "aether.exe"),
        };

        foreach (var name in new[] { Path.GetFileNameWithoutExtension(EngineProcessNames.Aether), "aether" })
        {
            Process[] found;
            try { found = Process.GetProcessesByName(name); }
            catch { continue; }

            foreach (var p in found)
            {
                try
                {
                    var image = WintunRouteApi.TryGetProcessPath(p.Id);
                    if (image is null ||
                        !ourImages.Any(our => string.Equals(image, our, StringComparison.OrdinalIgnoreCase)))
                    {
                        continue;
                    }
                    _logger.LogInformation("Stopping our Aether core (pid {Pid}) before replacing it", p.Id);
                    p.Kill(entireProcessTree: true);
                    p.WaitForExit(2000);
                }
                catch (Exception ex)
                {
                    _logger.LogDebug(ex, "Could not stop Aether pid {Pid}", p.Id);
                }
                finally
                {
                    try { p.Dispose(); } catch { }
                }
            }
        }
    }

    private static string FindAetherExecutable()
    {
        var localCached = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "Se7en", "aether", EngineProcessNames.Aether);
        if (File.Exists(localCached)) return localCached;

        var appResource = Path.Combine(AppContext.BaseDirectory, "Resources", "aether", "aether.exe");
        if (File.Exists(appResource)) return appResource;

        return "";
    }

    private static string FindTorExecutable()
    {
        var localCached = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "Se7en", "tor", EngineProcessNames.Tor);
        if (File.Exists(localCached)) return localCached;

        var appResource = Path.Combine(AppContext.BaseDirectory, "Resources", "tor", "tor.exe");
        if (File.Exists(appResource)) return appResource;

        return "";
    }

    private static string QueryExeVersion(string exePath, string arguments)
    {
        try
        {
            using var proc = new Process
            {
                StartInfo = new ProcessStartInfo
                {
                    FileName = exePath,
                    Arguments = arguments,
                    UseShellExecute = false,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    CreateNoWindow = true
                }
            };
            proc.Start();
            var output = proc.StandardOutput.ReadToEnd();
            proc.WaitForExit(2000);
            return output.Trim();
        }
        catch
        {
            return "";
        }
    }

    private static bool IsNewerVersion(string current, string latest)
    {
        if (string.IsNullOrWhiteSpace(latest)) return false;
        if (string.IsNullOrWhiteSpace(current)) return true;

        if (Version.TryParse(CleanVersion(current), out var cVer) &&
            Version.TryParse(CleanVersion(latest), out var lVer))
        {
            return lVer > cVer;
        }

        return !string.Equals(current.Trim(), latest.Trim(), StringComparison.OrdinalIgnoreCase);
    }

    private static string CleanVersion(string v)
    {
        var s = v.Trim().TrimStart('v', 'V');
        var dash = s.IndexOf('-');
        if (dash > 0) s = s.Substring(0, dash);
        return s;
    }

    private static string GetCoreDisplayName(string coreId) => coreId.ToLowerInvariant() switch
    {
        "aether" => "Aether (WARP / MASQUE)",
        "singbox" => "Sing-Box (Universal Engine)",
        "tor" => "Tor (Onion Routing)",
        "xray" => "Xray (V2Ray Platform)",
        _ => coreId
    };
}

File: Se7enFlutter/lib/core/services/core_update_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

class CoreUpdateInfo {
  const CoreUpdateInfo({
    required this.coreId,
    required this.displayName,
    required this.installedVersion,
    required this.latestVersion,
    required this.hasUpdate,
    required this.downloadUrl,
    required this.releaseNotes,
    required this.downloadSizeBytes,
  });

  final String coreId;
  final String displayName;
  final String installedVersion;
  final String latestVersion;
  final bool hasUpdate;
  final String downloadUrl;
  final String releaseNotes;
  final int downloadSizeBytes;
}

class CoreUpdateService {
  static final CoreUpdateService instance = CoreUpdateService._();
  CoreUpdateService._();

  final _httpClient = HttpClient()
    ..connectionTimeout = const Duration(seconds: 15)
    ..userAgent = 'Se7enPro-CoreUpdater/1.0';

  String getInstalledVersion(String coreId) {
    try {
      final id = coreId.toLowerCase();
      if (id == 'aether') {
        final path = _findAetherPath();
        if (path != null && File(path).existsSync()) {
          final ver = _queryExeVersion(path, '--version');
          if (ver != null && ver.isNotEmpty) {
            final parts = ver.split(' ');
            if (parts.length >= 2) return parts[1].trim();
            return ver.trim();
          }
        }
        return '1.7.0';
      } else if (id == 'tor') {
        final path = _findTorPath();
        if (path != null && File(path).existsSync()) {
          final ver = _queryExeVersion(path, '--version');
          if (ver != null && ver.isNotEmpty) {
            final lines = ver.split('\n');
            if (lines.isNotEmpty) {
              final first = lines.first.trim();
              final idx = first.toLowerCase().indexOf('version ');
              if (idx >= 0) {
                final rest = first.substring(idx + 8).trim();
                final space = rest.indexOf(' ');
                return space > 0 ? rest.substring(0, space) : rest;
              }
            }
          }
        }
        return '0.4.9.11';
      }
    } catch (_) {}
    return coreId.toLowerCase() == 'tor' ? '0.4.9.11' : '1.7.0';
  }

  Future<CoreUpdateInfo> checkForUpdate(String coreId) async {
    final id = coreId.toLowerCase();
    final installed = getInstalledVersion(id);

    if (id == 'tor') {
      await Future.delayed(const Duration(milliseconds: 400));
      return CoreUpdateInfo(
        coreId: 'tor',
        displayName: 'Tor (Onion Routing)',
        installedVersion: installed,
        latestVersion: installed,
        hasUpdate: false,
        downloadUrl: '',
        releaseNotes: 'Tor engine is running the latest bundled release.',
        downloadSizeBytes: 0,
      );
    }

    if (id != 'aether') {
      return CoreUpdateInfo(
        coreId: id,
        displayName: coreId,
        installedVersion: installed,
        latestVersion: installed,
        hasUpdate: false,
        downloadUrl: '',
        releaseNotes: '',
        downloadSizeBytes: 0,
      );
    }

    try {
      final uri = Uri.parse('https://api.github.com/repos/CluvexStudio/Aether/releases/latest');
      final request = await _httpClient.getUrl(uri);
      final response = await request.close();
      if (response.statusCode != 200) {
        throw HttpException('GitHub API returned ${response.statusCode}');
      }

      final body = await response.transform(utf8.decoder).join();
      final Map<String, dynamic> json = jsonDecode(body) as Map<String, dynamic>;

      final tagName = (json['tag_name'] as String? ?? '').trim();
      final latestVersion = tagName.replaceAll(RegExp(r'^[vV]'), '').trim();
      final releaseNotes = json['body'] as String? ?? '';

      String downloadUrl = '';
      int downloadSizeBytes = 0;

      final assets = json['assets'] as List<dynamic>? ?? [];
      for (final asset in assets) {
        final name = (asset['name'] as String? ?? '').toLowerCase();
        if (name == 'aether-windows-x86_64.zip') {
          downloadUrl = asset['browser_download_url'] as String? ?? '';
          downloadSizeBytes = (asset['size'] as num?)?.toInt() ?? 0;
          break;
        }
      }

      if (downloadUrl.isEmpty && tagName.isNotEmpty) {
        downloadUrl =
            'https://github.com/CluvexStudio/Aether/releases/download/$tagName/aether-windows-x86_64.zip';
      }

      final hasUpdate = _isNewerVersion(installed, latestVersion);

      return CoreUpdateInfo(
        coreId: 'aether',
        displayName: 'Aether (WARP / MASQUE)',
        installedVersion: installed,
        latestVersion: latestVersion.isEmpty ? installed : latestVersion,
        hasUpdate: hasUpdate,
        downloadUrl: downloadUrl,
        releaseNotes: releaseNotes,
        downloadSizeBytes: downloadSizeBytes,
      );
    } catch (e) {
      return CoreUpdateInfo(
        coreId: 'aether',
        displayName: 'Aether (WARP / MASQUE)',
        installedVersion: installed,
        latestVersion: installed,
        hasUpdate: false,
        downloadUrl: '',
        releaseNotes: 'Could not fetch remote version: $e',
        downloadSizeBytes: 0,
      );
    }
  }

  Future<bool> updateAether({
    required void Function(int progress, String status) onProgress,
  }) async {
    onProgress(5, 'Checking latest release…');
    final info = await checkForUpdate('aether');

    var downloadUrl = info.downloadUrl;
    if (downloadUrl.isEmpty) {
      downloadUrl =
          'https://github.com/CluvexStudio/Aether/releases/latest/download/aether-windows-x86_64.zip';
    }

    final tempDir = Directory(
        '${Directory.systemTemp.path}\\Se7en_Update_${DateTime.now().millisecondsSinceEpoch}');
    await tempDir.create(recursive: true);

    try {
      final zipFile = File('${tempDir.path}\\aether-windows-x86_64.zip');
      onProgress(10, 'Starting download…');

      final uri = Uri.parse(downloadUrl);
      final request = await _httpClient.getUrl(uri);
      final response = await request.close();
      if (response.statusCode != 200 && response.statusCode != 302) {
        throw HttpException('Download failed: HTTP ${response.statusCode}');
      }

      final totalBytes = response.contentLength > 0
          ? response.contentLength
          : (info.downloadSizeBytes > 0 ? info.downloadSizeBytes : 15000000);
      var receivedBytes = 0;
      final sink = zipFile.openWrite();

      await for (final chunk in response) {
        sink.add(chunk);
        receivedBytes += chunk.length;
        final pct = (10 + ((receivedBytes / totalBytes) * 70).clamp(0, 70)).toInt();
        onProgress(pct, 'Downloading: $pct%');
      }
      await sink.flush();
      await sink.close();

      onProgress(85, 'Extracting archive…');
      final extractDir = Directory('${tempDir.path}\\extracted');
      await extractDir.create(recursive: true);

      final extractRes = await Process.run('tar', ['-xf', zipFile.path, '-C', extractDir.path]);
      if (extractRes.exitCode != 0) {
        await Process.run('powershell', [
          '-NoProfile',
          '-Command',
          "Expand-Archive -LiteralPath '${zipFile.path}' -DestinationPath '${extractDir.path}' -Force"
        ]);
      }

      File? extractedExe;
      await for (final f in extractDir.list(recursive: true)) {
        if (f is File && f.path.toLowerCase().endsWith('aether.exe')) {
          extractedExe = f;
          break;
        }
      }

      if (extractedExe == null || !extractedExe.existsSync()) {
        throw const FileSystemException('Extracted package did not contain aether.exe');
      }

      onProgress(92, 'Installing binary…');
      await Process.run('taskkill', ['/F', '/IM', 'aether.exe', '/T'])
          .catchError((_) => ProcessResult(0, 0, '', ''));
      await Process.run('taskkill', ['/F', '/IM', 'Se7enPro.Aether.exe', '/T'])
          .catchError((_) => ProcessResult(0, 0, '', ''));

      final appData = Platform.environment['LOCALAPPDATA'] ?? '';
      if (appData.isNotEmpty) {
        final targetCached = File('$appData\\Se7en\\aether\\aether.exe');
        final targetNamed = File('$appData\\Se7en\\aether\\Se7enPro.Aether.exe');
        targetCached.parent.createSync(recursive: true);
        try {
          if (targetCached.existsSync()) targetCached.deleteSync();
        } catch (_) {}
        try {
          if (targetNamed.existsSync()) targetNamed.deleteSync();
        } catch (_) {}
        extractedExe.copySync(targetCached.path);
        extractedExe.copySync(targetNamed.path);
      }

      try {
        final exeDir = File(Platform.resolvedExecutable).parent.path;
        final resFile = File('$exeDir\\Resources\\aether\\aether.exe');
        if (resFile.parent.existsSync()) {
          extractedExe.copySync(resFile.path);
        }
      } catch (_) {}

      try {
        File? extractedPt;
        await for (final f in extractDir.list(recursive: true)) {
          if (f is File && f.path.toLowerCase().endsWith('lyrebird.exe')) {
            extractedPt = f;
            break;
          }
        }
        if (extractedPt != null) {
          if (appData.isNotEmpty) {
            final ptDir = Directory('$appData\\Se7en\\aether\\pt');
            ptDir.createSync(recursive: true);
            extractedPt.copySync('${ptDir.path}\\lyrebird.exe');
          }
          try {
            final exeDir = File(Platform.resolvedExecutable).parent.path;
            final resPt = File('$exeDir\\Resources\\aether\\pt\\lyrebird.exe');
            resPt.parent.createSync(recursive: true);
            extractedPt.copySync(resPt.path);
          } catch (_) {}
        }
      } catch (_) {}

      onProgress(100, 'Aether core successfully installed!');
      return true;
    } finally {
      try {
        if (tempDir.existsSync()) {
          tempDir.deleteSync(recursive: true);
        }
      } catch (_) {}
    }
  }

  bool _isNewerVersion(String current, String latest) {
    try {
      final cParts = current.split('.').map((p) => int.tryParse(p) ?? 0).toList();
      final lParts = latest.split('.').map((p) => int.tryParse(p) ?? 0).toList();
      final len = cParts.length > lParts.length ? cParts.length : lParts.length;
      while (cParts.length < len) {
        cParts.add(0);
      }
      while (lParts.length < len) {
        lParts.add(0);
      }
      for (var i = 0; i < len; i++) {
        if (lParts[i] > cParts[i]) return true;
        if (lParts[i] < cParts[i]) return false;
      }
    } catch (_) {}
    return false;
  }

  String? _findAetherPath() {
    final appData = Platform.environment['LOCALAPPDATA'];
    if (appData != null) {
      final cached = File('$appData\\Se7en\\aether\\aether.exe');
      if (cached.existsSync()) return cached.path;
      final cachedNamed = File('$appData\\Se7en\\aether\\Se7enPro.Aether.exe');
      if (cachedNamed.existsSync()) return cachedNamed.path;
    }
    try {
      final exeDir = File(Platform.resolvedExecutable).parent.path;
      final res = File('$exeDir\\Resources\\aether\\aether.exe');
      if (res.existsSync()) return res.path;
    } catch (_) {}
    final candidate = File('${Directory.current.path}\\Resources\\aether\\aether.exe');
    if (candidate.existsSync()) return candidate.path;
    return null;
  }

  String? _findTorPath() {
    final candidate = File('${Directory.current.path}\\Resources\\tor\\tor.exe');
    if (candidate.existsSync()) return candidate.path;
    return null;
  }

  String? _queryExeVersion(String exePath, String arg) {
    try {
      final res = Process.runSync(exePath, [arg]);
      if (res.exitCode == 0) {
        return (res.stdout as String? ?? '').trim();
      }
    } catch (_) {}
    return null;
  }
}
